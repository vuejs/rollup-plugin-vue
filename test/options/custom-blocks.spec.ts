import vue, { VuePluginOptions } from '../../src'
import { pluginInline } from '../setup/plugins'
import { rollup } from 'rollup'
function pluginText() {
  return {
    name: 'text',
    transform(source: string, id: string) {
      if (/\.(md|txt)$/.test(id)) {
        return `export default ${JSON.stringify(source.trim())}`
      }
    },
  }
}

describe('customBlocks', () => {
  async function setup(options?: Partial<VuePluginOptions>) {
    return rollup({
      input: '/entry.vue',
      plugins: [
        pluginInline(
          '/entry.vue',
          `
        <template>
        <div>Hello, world</div>
        </template>
        <custom>
        // My Custom Block
        </custom>
        <docs>
        // My Docs Block
        </docs>
      `
        ),
        pluginText(),
        vue({
          ...options,
          defaultLang: {
            docs: 'md',
            custom: 'txt',
          },
          normalizer: 'vue-runtime-helpers/dist/normalize-component.mjs',
        }),
      ],
    })
      .then(bundle => bundle.generate({ format: 'es' }))
      .then(generated => generated.output[0])
  }

  it('default', async () => {
    const { code } = await setup()

    expect(code).not.toEqual(expect.stringContaining('My Custom Block'))
    expect(code).not.toEqual(expect.stringContaining('My Docs Block'))
  })

  it('array of tags', async () => {
    const { code } = await setup({
      customBlocks: ['custom'],
    })

    expect(code).toEqual(expect.stringContaining('My Custom Block'))
    expect(code).not.toEqual(expect.stringContaining('My Docs Block'))
  })
  it('negative array of tags', async () => {
    const { code } = await setup({
      customBlocks: ['*', '!custom'],
    })

    expect(code).not.toEqual(expect.stringContaining('My Custom Block'))
    expect(code).toEqual(expect.stringContaining('My Docs Block'))
  })
  it('function', async () => {
    const { code } = await setup({
      customBlocks(tag) {
        return tag === 'custom'
      },
    })

    expect(code).toEqual(expect.stringContaining('My Custom Block'))
    expect(code).not.toEqual(expect.stringContaining('My Docs Block'))
  })

  it('transform', async () => {
    const { code } = await setup({
      customBlocks: ['docs'],
    })

    expect(code).toEqual(expect.stringContaining('__custom_block_1__(__vue_component__)'))
    expect(code).toMatchSnapshot()
  })
})
