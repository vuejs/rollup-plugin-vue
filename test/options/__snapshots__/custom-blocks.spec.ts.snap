// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customBlocks transform 1`] = `
"var __custom_block_1__ = \\"// My Docs Block\\";

/* script */

/* template */
var __vue_render__ = function() {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(\\"div\\", [_vm._v(\\"Hello, world\\")])
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* component normalizer */
  const __vue_normalize__ = vue-runtime-helpers/dist/normalize-component.mjs;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = __vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    {},
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );
__custom_block_1__(__vue_component__);

export default __vue_component__;
"
`;
